hibernate-core/src/test/java/org/hibernate/orm/test/mapping/identifier/composite/CompositEntityUpdateTest.java

```diff
@@ -0,0 +1,212 @@
+import org.hibernate.testing.orm.junit.Setting;
+import org.junit.jupiter.api.Test;
+
+import java.io.Serial;
+import java.io.Serializable;
+import java.time.Instant;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@DomainModel(
+		annotatedClasses = {
+				CompositEntityUpdateTest.TimeScalePkOidDatum.class,
+				CompositEntityUpdateTest.TimeScalePkOidDatumEmbed.class,
+				CompositEntityUpdateTest.TestEntity.class,
+				CompositEntityUpdateTest.TestEntityEmbed.class,
+		}
+)
+@ServiceRegistry(
+		settings = {
+				// For your own convenience to see generated queries:
+				@Setting(name = AvailableSettings.SHOW_SQL, value = "true"),
+				@Setting(name = AvailableSettings.FORMAT_SQL, value = "true"),
+				// @Setting( name = AvailableSettings.GENERATE_STATISTICS, value = "true" ),
+		}
+)
+@SessionFactory
+@Jira("HHH-18721")
+public class CompositEntityUpdateTest {
+
+	@Test
+	void hhh18721TestIdClass(SessionFactoryScope scope) {
+		scope.inTransaction( em -> {
+			TestEntity e = new TestEntity("yyy", "foo");
+			em.persist(e);
+			em.flush();
+			assertTrue(em.contains(e));
+			assertEquals(e.getData(), "foo");
+
+			e.updateData("bar");
+			em.flush();
+			assertEquals(e.getData(), "bar");
+		} );
+	}
+
+	@Test
+	void hhh18721TestEmbedded(SessionFactoryScope scope) {
+		scope.inTransaction( em -> {
+			TestEntityEmbed e = new TestEntityEmbed("yyy", "foo");
+			em.persist(e);
+			em.flush();
+			assertTrue(em.contains(e));
+			assertEquals(e.getData(), "foo");
+
+			e.updateData("bar");
+			em.flush();
+			assertEquals(e.getData(), "bar");
+		} );
+	}
+
+	/**
+	 * Composite primary key class for oid and datum
+	 */
+	public static class TimeScalePkOidDatum implements Serializable {
+
+		@Serial
+		private static final long serialVersionUID = 1L;
+
+		private String oid;
+		private Instant datum;
+
+		public TimeScalePkOidDatum() {
+			// Default constructor
+		}
+
+		public TimeScalePkOidDatum(String oid, Instant datum) {
+			this.oid = oid;
+			this.datum = datum;
+		}
+
+		public String oid() {
+			return this.oid;
+		}
+
+		public Instant datum() {
+			return this.datum;
+		}
+	}
+
+	@Embeddable
+	public static class TimeScalePkOidDatumEmbed implements Serializable {
+
+		@Serial
+		private static final long serialVersionUID = 1L;
+
+		private String oid;
+		private Instant datum;
+
+		public TimeScalePkOidDatumEmbed() {
+			// Default constructor
+		}
+
+		public TimeScalePkOidDatumEmbed(String oid, Instant datum) {
+			this.oid = oid;
+			this.datum = datum;
+		}
+
+		public String oid() {
+			return this.oid;
+		}
+
+		public Instant datum() {
+			return this.datum;
+		}
+	}
+
+	@Entity
+	@IdClass(TimeScalePkOidDatum.class)
+	public static class TestEntity {
+
+		@Id
+		@Column(length = 32, nullable = false)
+		private String oid = null;
+
+		@Id
+		@Column(columnDefinition = "TIMESTAMP WITH TIME ZONE", nullable = false)
```

Comment:
Our generators can do the rounding/truncation for you:

```suggestion
		@CreationTimestamp
		@FractionalSeconds(6)
		@Column(nullable = false)
```